### ISSUE TITLE
Mod version management

### ISSUE THREAD

**Issue by kianzarrin** [krzychu124/Cities-Skylines-Traffic-Manager-President-Edition#678]
In #649 I need to get the  version of the mod based on which I can decide whether CSUT_Toolbox should patch TMPE or not. we may use this:
```c#
TrafficManagerMod::public const string VERSION = "11.0";
```
see 
https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/issues/649#issuecomment-583849196
https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/issues/649#issuecomment-583856602

Also it would be nice to generate the version from AssemblyInfo.cs so we have a central place for doing all this: https://stackoverflow.com/questions/15126113/getting-assembly-version-from-assemblyinfo-cs

**Comment by kianzarrin** [comment]
To share assembly version across projects:
https://weblogs.asp.net/ashishnjain/sharing-assembly-version-across-projects-in-a-solution

also  from https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/issues/649#issuecomment-583881299
>
>Just realised VERSION isn't part of IUserMod so we can change it to any format we want
@kianzarrin Change it to number but provide some way to include string bolt ons in the stuff that gets shown to user (eg. 11.1-beta1 or 11.0-hotfix)
I prefer to use Version class. 
```c#
// usage example
new Version(11.1,blah,blah)
version.ToString(2) //major and minor only
version1 > version2 // it has internal comparison
```
>And I will get `Version` from AssemblyInfo.cs:
```C#
       public static string VersionString => ModVersion.ToString(2);
       public static Version ModVersion => typeof(TrafficManagerMod).Assembly.GetName().Version;
```
>Hopefully no one has used `TrafficManagerMod::VERSION` externally.

**Comment by kianzarrin** [comment]
Now there is only once place to change mod version and all assembly versions:
`SharedAssemblyInfo.cs`
So tidy ... I love it!

Here is a proof that it works:
![image](https://user-images.githubusercontent.com/26344691/74109259-86f51480-4b8a-11ea-8b04-09facb314ed0.png)
![Screenshot (441)](https://user-images.githubusercontent.com/26344691/74109274-a2601f80-4b8a-11ea-9779-cc21a34eb740.png)

**Comment by krzychu124** [comment]
Also we should update API project interfaces and make strong version instead of `[1.0.*]` 
Probably then we'll need to log changes in separate section of `Changelog` , maybe even export whole project as a new repository and release each version separately (it still could be linked as submodule in main repo). 
Any mod dev could download API lib and import it in his mod source code and relay on provided interfaces when he/she want to make changes to `TM:PE` settings. 
Still thinking how to expose manager instances reliably (return interface instance instead of using it directly) :/

**Comment by kianzarrin** [comment]
> Also we should update API project interfaces 

you mean to provide an interface to get version? I never understood how API works. I used TrafficMangerMod.dll in my mod to access pedestrian crossing rules.

> and make strong version instead of [1.0.*]

I don't suggest to use strong version because hot reload will not work. in API we can `return new Version(version.major, version.minor)` and thus hiding the build and revision for interface. I don't see the why is it important to do so though.

> Probably then we'll need to log changes in separate section of Changelog , maybe even export whole project as a new repository and release each version separately (it still could be linked as submodule in main repo).
Any mod dev could download API lib and import it in his mod source code and relay on provided interfaces when he/she want to make changes to TM:PE settings. Still thinking how to expose manager instances reliably (return interface instance instead of using it directly) :/

Interesting stuff. I never knew that's how API is meant to be used. That's sounds like a complicated task that requires a lot of learning on my part. it's worth putting into a new issue. if we divide tasks into small steps, they would be easily achievable.

**Comment by aubergine10** [comment]
If someone can provide brain dump of basics of using the API I'll tidy it in to markdown and we can add some docs to [the relevant folder](https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/tree/master/TLM/TMPE.API) and maybe link from contributing guide or something.

I assume it just spits out a DLL that devs can reference from their projects to interface to TMPE at runtime?

**Comment by krzychu124** [comment]
The missing part is: reliable source of instances, since Managers in TM:PE are static. 
Probably we would need class in API which would act as container to provide instances or we could use some IoC container like [Unity](https://github.com/unitycontainer/container) 
Quick example of usage: 
```cs
public class InstanceProvider {
   public ITrafficPriorityManager GetPriorityManager() {
      return TrafficPriorityManager.instance;
   }
}
```
then other mods could use it as follows (it's only example):
```cs
bool hasNode = InstanceProvider.GetPriorityManager().HasNodePrioritySign(1234);
//TrafficPriorityManager has declaration of bool HasNodePrioritySign(ushort nodeId);
```
I think it's pretty standard Plug-in architecture.
Having instance provider in API you get ability to modify everything in main project without breaking other mods. You can change everything, class/property names, types, the only thing that should stay intact is API call. It should always return instance compatible with interface. In my opinion strong versioning is important for maintaining compatibility. You can add more declarations but you shouldn't change existing, since it will cause breaking change. Say mod x use API v1 and wants to call some method, it shouldn't break anything when we update API to v2 with new method declarations unless we remove/change any old declarations. 
The only thing which might cause issues I can imagine is hot-reload, but we would need to test out it.

I suggest reading about `extern` keyword, as some way to resolve runtime versioning issues :)

_I'm not C# dev so I might be completely wrong with this approach. I'm open on other ideas._

**Comment by aubergine10** [comment]
Suggest start a new issue for tracking updating the API.
